### PR TITLE
Driver interface

### PR DESCRIPTION
**Pull request by tjs-intel**
This introduces a new `Driver` interface. A correct `Driver` implementation has at least a single method `request`:
```python
def request(self, method: str, path: list[str], data: Optional[Any], headers: Optional[dict[str, str]]) -> Any: ...
```

A SigOpt `Connection` can be initialized with any driver implementation by providing the driver class as the `driver` keyword argument:
```python
Connection(driver=CustomDriver)
```

`"http"` and `"lite"` are also acceptable arguments for the `driver` keyword argument. `"http"` will initialize a `RequestDriver` which calls out to `os.environ("SIGOPT_API_URL", "https://api.sigopt.com")`. `"lite"` will initialize a `LocalDriver` provided by the `sigoptlite` package, which performs optimization locally without any communications to a SigOpt server. The `"lite"` driver implementation will be available via `pip install 'sigopt[lite]'` when the `sigoptlite` module is released.

The default `driver` keyword argument is `"http"` to preserve compatibility with existing code.

### CODE REVIEW

File: sigopt/interface.py [hunk 1]
```diff
@@ -302,98 +275,72 @@ def __init__(self, requestor, api_url=None, user_agent=None, verify_ssl_certs=No
...
+  try:
+    from sigoptlite import LocalDriver  # pylint: disable=import-error
+  except ModuleNotFoundError as mnfe:
+    raise ModuleNotFoundError(
+      "SigOpt Lite is not installed. It can be installed with the following command: `pip install 'sigopt[lite]'`"
+    ) from mnfe
+  return LocalDriver(*args, **kwargs)
+
+DRIVER_KEY_HTTP = "http"
+DRIVER_KEY_LITE = "lite"
```
Comment (simonh-intel) [inline sigopt/interface.py:314]: Research I think wants to use "local" will confirm at meeting today

File: sigopt/request_driver.py [hunk 2]
```diff
@@ -81,18 +120,24 @@ def _request(self, method, url, params, json, headers, user_agent):
...
       jitter=backoff.full_jitter,
     )
-    response = retry(self._request)(method, url, params, json, headers, user_agent)
+    response = retry(self._request)(method, url, params, json, headers)
     return self._handle_response(response)
 
-  def _with_default_headers(self, headers, user_agent):
-    user_agent_str = user_agent or 'sigopt-python/{0}'.format(VERSION)
+  def _with_default_headers(self, headers):
+    user_agent_str = 'sigopt-python/{0}'.format(VERSION)
```
Comment (dwanderson-intel) [inline sigopt/request_driver.py:140]: tidbit - no longer python2, so we can probably just `f"sigopt-python/{VERSION}"`?

**Review by dwanderson-intel — APPROVED** [review]
non-blocking notes

File: sigopt/interface.py (hunk 1 above)
Comment (simonh-intel) [reply sigopt/interface.py:314]: thanks yeah they are onboard with "lite"

File: sigopt/request_driver.py [hunk 3]
```diff
@@ -2,61 +2,100 @@
...
     timeout=DEFAULT_HTTP_TIMEOUT,
     client_ssl_certs=None,
     session=None,
+    api_url=None,
   ):
-    self._set_auth(user, password)
-    self.default_headers = headers or {}
-    self.verify_ssl_certs = verify_ssl_certs
+    if client_token is None:
+      client_token = client_token or os.environ.get("SIGOPT_API_TOKEN", config.api_token)
```
[the remark this answers is not in the document]
Comment (simonh-intel) [reply sigopt/request_driver.py:47]: falsy things i.e. empty string would still get caught by the or - not sure if relevant

[the remark this answers is not in the document]
Comment (tjs-intel) [reply sigopt/request_driver.py:47]: True, but I don't think we need this distinction here

**Pull request merged**